lib/dust.js

```diff
@@ -743,37 +774,26 @@
   };
 
   Chunk.prototype.partial = function(elem, context, params) {
-    var partialContext;
-    //put the params context second to match what section does. {.} matches the current context without parameters
-    // start with an empty context
-    partialContext = dust.makeBase(context.global);
-    partialContext.blocks = context.blocks;
-    if (context.stack && context.stack.tail){
-      // grab the stack(tail) off of the previous context if we have it
-      partialContext.stack = context.stack.tail;
-    }
-    if (params){
-      //put params on
-      partialContext = partialContext.push(params);
-    }
+    var head;
 
-    if(typeof elem === 'string') {
-      partialContext.templateName = elem;
+    if (!dust.isEmptyObject(params)) {
+      context = context.clone();
+      head = context.pop();
+      context = context.push(params)
+                       .push(head);
     }
 
-    //reattach the head
-    partialContext = partialContext.push(context.stack.head);
-
-    var partialChunk;
-    if (typeof elem === 'function') {
-      partialChunk = this.capture(elem, partialContext, function(name, chunk) {
-        partialContext.templateName = partialContext.templateName || name;
-        dust.load(name, chunk, partialContext).end();
+    if (elem.__dustBody) {
+      // The eventual result of evaluating `elem` is a partial name
+      // Load the partial after getting its name and end the async chunk
+      return this.capture(elem, context, function(name, chunk) {
+        context.templateName = name;
```

Comment:
The templateName is stored outside of the stack, so there shouldn't be any effect by attaching it at an arbitrary time as it's not part of the stack traversal.